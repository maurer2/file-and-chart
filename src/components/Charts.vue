<template>
  <div class="chart">
    <h2>Charts</h2>

    <pre>{{ chartLabels }}</pre>
    <h3>Bar chart</h3>
    <canvas ref="barchart" />
  </div>
</template>

<script>
import Chart from 'chart.js';

export default {
  name: 'Charts',
  props: {
    chartData: Array,
  },
  data() {
    return {
    }
  },
  computed: {
    chartLabels() {
      const names = this.chartData.map((tableRow) => tableRow.name);

      return names;
    },
    chartValues() {
      const age = this.chartData.map((tableRow) => tableRow.age);

      return age;
    }
  },
  methods: {},
  mounted() {
    const barChart = new Chart(this.$refs.barchart, {
      type: 'horizontalBar',
      options: {
        legend: {
          display: false
        },
      },
      data: {
        datasets: [{
          label: 'Cat ages',
          type: 'horizontalBar',
          data: this.chartValues,
        }],
        labels: this.chartLabels,
      },
    });
  }
};
</script>

<style scoped lang="scss">

</style>
