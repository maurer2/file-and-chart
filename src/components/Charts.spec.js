import { shallowMount } from '@vue/test-utils';
import 'jest-canvas-mock';
import Charts from '@/components/Charts.vue';

const chartData = [
  { name: 'Bob', age: '15' },
  { name: 'Boots', age: '21' },
  { name: 'Cat', age: '32' },
  { name: 'Felix', age: '44' },
  { name: 'Garfield', age: '51' },
];

describe('Charts', () => {
  let wrapper;

  const mockedCreateBarChart = jest.fn();
  const mockedCreateCircleChart = jest.fn();

  beforeEach(() => {
    wrapper = shallowMount(Charts, {
      propsData: {
        chartData,
      },
      attachToDocument: true,
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('Charts snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Charts contains all elements and child elements', () => {
    expect(wrapper.find('.charts').exists()).toBe(true);
    expect(wrapper.find('.chart-container:nth-child(1)').exists()).toBe(true);
    expect(wrapper.find('.chart-container:nth-child(2)').exists()).toBe(true);
    expect(wrapper.find('.barchart').exists()).toBe(true);
    expect(wrapper.find('.circlechart').exists()).toBe(true);
  });

  test('Getter chartLabels returns array of strings', () => {
    expect(wrapper.vm.chartLabels).toEqual(['Bob', 'Boots', 'Cat', 'Felix', 'Garfield']);

    wrapper.setProps({ chartData: [] });
    expect(wrapper.vm.chartLabels).toEqual([]);
  });

  test('Getter chartValues returns array of numbers', () => {
    expect(wrapper.vm.chartValues).toEqual(['15', '21', '32', '44', '51']);

    wrapper.setProps({ chartData: [] });
    expect(wrapper.vm.chartValues).toEqual([]);
  });

  test('Charts barChart and circleChart are defined', () => {
    expect(wrapper.vm.barChart).toBeDefined();
    expect(wrapper.vm.circleChart).toBeDefined();
  });

  test('Cart new props trigger chart creation', () => {
    wrapper.setMethods({ createBarChart: mockedCreateBarChart });
    wrapper.setMethods({ createCircleChart: mockedCreateCircleChart });

    wrapper.setProps({ chartData: chartData.slice(1) });

    expect(mockedCreateBarChart).toBeCalled();
    expect(mockedCreateCircleChart).toBeCalled();
  });
});
