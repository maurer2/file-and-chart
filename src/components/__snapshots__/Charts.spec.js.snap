// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Charts Charts snapshot 1`] = `
<div class="charts">
  <div class="chart-container">
    <h3>Bar chart</h3> <canvas class="barchart chartjs-render-monitor" style="display: block;" height="0" width="0"></canvas>
  </div>
  <div class="chart-container">
    <h3>Circle chart</h3> <canvas class="circlechart chartjs-render-monitor" style="display: block;" height="0" width="0"></canvas>
  </div>
</div>
`;
